ruby 2.4.3, ruby@2.3 2.3.6, ruby@2.2 2.2.9 & rubygems 2.7.3

lembacon opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

ilovezfs commented on Formula/ruby.rb:
```diff
@@ -36,8 +35,8 @@ class Ruby < Formula
   resource "rubygems" do
-    url "https://rubygems.org/rubygems/rubygems-2.6.14.tgz"
-    sha256 "406a45d258707f52241843e9c7902bbdcf00e7edc3e88cdb79c46659b47851ec"
+    url "https://rubygems.org/rubygems/rubygems-2.7.3.tgz"
```
@DomT4 thoughts on bumping this? My inclination is to leave it alone at 2.6.14

DomT4 replied:
There are some pretty nice changes in the new versions of `gem`, so if it works across all our versioned formulae I'd be tempted to bump, but sticking my maintainer hat back on for a moment I'd certainly agree there's nothing that _needs_ to be pushed out to users. `gem update --system` is quite a well-known command, available to anyone who cares particularly. I lean towards updating, but it's only a gentle lean.

ilovezfs replied:
@lembacon OK let's try updating gem in ${all_the_rubies} and revision bumping them, revision bumping all the rev deps of ruby since things seem to have decided to link based on the patch-level version of the library :see_no_evil:

DomT4 replied:
Some of the other versioned formulae (`ruby@2.2` and `ruby@2.3`) have new releases (I think?) so they can be updated rather than revision bumped at least. Less hideousness 🙈.

DomT4 commented:
Thoughts:

* `ruby@1.8` may be a lost cause, unless we want to vendor in a newer `psych` for that formula as well, providing the newer versions of `psych` even support 1.8. IMO `ruby@1.8` can live as-is until the next known vuln in `gem` and then we might have to give it the boot.
* Anything using Homebrew's version of `ruby` shouldn't need `gem install bundler` steps in `test do` as there's a perfectly good vendored `bundler` with `gem`. I'd try simply removing those lines for the two tests that failed.

lembacon commented:
> `ruby@1.8` may be a lost cause, unless we want to vendor in a newer `psych` for that formula as well, providing the newer versions of `psych` even support 1.8. IMO `ruby@1.8` can live as-is until the next known vuln in `gem` and then we might have to give it the boot.

Removed the RubyGems version bump for `ruby@1.8`, and added a comment in the formula to clarify this.

lembacon commented:
> Anything using Homebrew's version of `ruby` shouldn't need `gem install bundler` steps in `test do` as there's a perfectly good vendored `bundler` with `gem`. I'd try simply removing those lines for the two tests that failed.

CI failures for formulae `bettercap` and `wpscan` do not seem to be related to the aforementioned `test do` block, however the failures of them are identical, both are due to `system libexec/"bin/bundle", "install", "--without", "test"` in the `def install` method where ``/usr/local/Cellar/ruby/2.4.3/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:59:in `require': cannot load such file -- bundler (LoadError)``.

@DomT4 Do you have any ideas regarding this failure? 🤨

ilovezfs commented:
what CI failures are we talking about

lembacon commented:
@ilovezfs This one https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/14802/version=high_sierra/console

ilovezfs commented:
I think you need this:
```
diff --git a/Formula/ruby.rb b/Formula/ruby.rb
index 64da5a42d6..1ed931aaad 100644
--- a/Formula/ruby.rb
+++ b/Formula/ruby.rb
@@ -117,10 +117,9 @@ class Ruby < Formula
       rm_f bin/"gem#{program_suffix}"
 
       # Drop in the new version.
-      (rg_in/"rubygems").install Dir[buildpath/"vendor_gem/lib/rubygems/*"]
-      rg_in.install buildpath/"vendor_gem/lib/rubygems.rb"
-      rg_in.install buildpath/"vendor_gem/lib/ubygems.rb"
+      rg_in.install Dir[buildpath/"vendor_gem/lib/*"]
       bin.install buildpath/"vendor_gem/bin/gem" => "gem#{program_suffix}"
+      bin.install buildpath/"vendor_gem/bin/bundle" => "bundle#{program_suffix}"
     end
   end
 
```
Files are missing.

lembacon commented:
Thanks @ilovezfs! Applied to all rubies with `RubyGems` upgraded to 2.7.3 (i.e. except `ruby@1.8`).

ilovezfs commented:
You're welcome. I suspect the
```
bin.install buildpath/"vendor_gem/bin/bundle" => "bundle#{program_suffix}"
```
may cause some link overwrite issues but let's at least see if we can get a green CI run.

ilovezfs commented:
@BrewTestBot test this please

ilovezfs commented:
Currently if you `gem update --system` it actually installs its own vendored `bundle` in `/usr/local/Cellar/ruby/2.4.2_1/bin/bundle`, which isn't symlinked into `/usr/local/bin` since you're running `gem update --system` after ruby has already been installed and linked.

But then if you `brew unlink ruby && brew link ruby`, it does symlink `bundle` into `/usr/local/bin`. So it seems kind of weird that we wouldn't symlink `bundle` in ourselves from the get-go.

That said, it will likely cause support issues since people won't be expecting to have to run `brew link --overwrite ruby` if they happen to have `bundler` installed, which is likely given that we haven't been installing it.

@DomT4 any thoughts?

DomT4 commented:
> @DomT4 any thoughts?

Difficult choice on this. IMO since it ships with `gem` and we ship `gem` the expected behaviour from a user point of view would be that we ship `bundle` as well, and I think there are reasonable and solid arguments for doing so, but yeah, it's not ideal that a chunk of users will likely have to spot & manually run `brew link --overwrite ruby`. I lean towards thinking we should do it anyway, but the linkage issue could be deeply annoying.

ilovezfs commented:
I'm thinking given that when there's a conflict with `/usr/local/bin/bundle`, ruby installs unlinked _and that unlinked state now persists across upgrades_ (arguably a `brew` bug since a failed brew link step during brew install actually isn't the same thing as the user explicitly requesting `brew unlink foo`) we may need to err on the side of caution here.

ilovezfs commented:
```
iMac-TMP:homebrew-core joe$ brew install ruby
==> Downloading https://homebrew.bintray.com/bottles/ruby-2.4.1_2.el_capitan.bottle.tar.gz
Already downloaded: /Users/joe/Library/Caches/Homebrew/ruby-2.4.1_2.el_capitan.bottle.tar.gz
==> Pouring ruby-2.4.1_2.el_capitan.bottle.tar.gz
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /usr/local
Could not symlink bin/rake
Target /usr/local/bin/rake
already exists. You may want to remove it:
  rm '/usr/local/bin/rake'

To force the link and overwrite all conflicting files:
  brew link --overwrite ruby

To list all files that would be deleted:
  brew link --overwrite --dry-run ruby

Possible conflicting files are:
/usr/local/bin/rake
==> Caveats
Emacs Lisp files have been installed to:
  /usr/local/share/emacs/site-lisp/ruby
==> Summary
🍺  /usr/local/Cellar/ruby/2.4.1_2: 1,191 files, 15.5MB
iMac-TMP:homebrew-core joe$ hash -r
iMac-TMP:homebrew-core joe$ which ruby
/usr/bin/ruby
iMac-TMP:homebrew-core joe$ git checkout origin/master -- Formula/ruby.rb 
iMac-TMP:homebrew-core joe$ brew upgrade ruby
==> Upgrading 1 outdated package, with result:
ruby 2.4.2_1
==> Upgrading ruby 
==> Downloading https://homebrew.bintray.com/bottles/ruby-2.4.2_1.el_capitan.bottle.tar.gz
Already downloaded: /Users/joe/Library/Caches/Homebrew/ruby-2.4.2_1.el_capitan.bottle.tar.gz
==> Pouring ruby-2.4.2_1.el_capitan.bottle.tar.gz
==> Caveats
Emacs Lisp files have been installed to:
  /usr/local/share/emacs/site-lisp/ruby
==> Summary
🍺  /usr/local/Cellar/ruby/2.4.2_1: 3,042 files, 16.3MB
iMac-TMP:homebrew-core joe$ hash -r
iMac-TMP:homebrew-core joe$ which ruby
/usr/bin/ruby
```
Notice no warning about the conflict the second time.

DomT4 commented:
> (arguably a brew bug since a failed brew link step during brew install actually isn't the same thing as the user explicitly requesting brew unlink foo)

Yes, this should absolutely be fixed first. I'm happy for us to leave it for now given these Ruby releases give some CVEs. We can revisit this later easily enough.

lembacon commented:
Did another force push to resolve conflicts against `Formula/vim.rb`.

ilovezfs commented:
Thanks @lembacon! Shipped. :gem: :ship:
